src/ui/base-page-storyEdit.js

```diff
@@ -274,50 +364,73 @@ https://raw.githubusercontent.com/fluid-project/sjrk-story-telling/main/LICENSE.
-     * @param {Component} storyEditor - an instance of `sjrk.storyTelling.ui.storyEditor`
-     * @param {String|String[]} storyPath - the model path to load the story to
-     * @param {String} sourceName - the name of the Infusion change source for this update
+     * @param {Component} storyEdit - an instance of `sjrk.storyTelling.base.page.storyEdit`
      */
-    sjrk.storyTelling.base.page.storyEdit.loadStoryFromAutosave = function (storyAutosaveKey, storyEditor, storyPath, sourceName) {
+    sjrk.storyTelling.base.page.storyEdit.initializeStory = function (storyAutosaveKey, storyEdit) {
         try {
             // localStorage can only store string values
-            var savedStory = JSON.parse(window.localStorage.getItem(storyAutosaveKey));
-
-            if (savedStory) {
-                // media block blob URLs are no longer valid/reliable after a page reload
-                sjrk.storyTelling.base.page.storyEdit.clearMediaBlockUrls(savedStory.content);
+            var savedStoryData = JSON.parse(window.localStorage.getItem(storyAutosaveKey));
 
-                storyEditor.story.applier.change(storyPath, savedStory, null, sourceName);
-
-                // build the storyEditor blockUIs from the story content array
-                storyEditor.blockManager.createBlocksFromData(savedStory.content);
+            if (savedStoryData) {
+                // a story was loaded from autosave, update the current story
+                storyEdit.loadStoryContent(savedStoryData);
             } else {
-                fluid.log(fluid.logLevel.WARN, "Story load aborted, no autosaved story was found");
+                // there's no autosaved story, create a new unpublished story
+                storyEdit.createNewStoryOnServer();
             }
         } catch (ex) {
-            fluid.log(fluid.logLevel.WARN, "An error occurred when loading", ex);
+            fluid.log(fluid.logLevel.WARN, "An error occurred while initializing story", ex);
         }
     };
 
     /**
-     * Removes the URL values of all media blocks within a collection of blocks
+     * Updates the storyEditor's story and blockManager with saved story data
      *
-     * @param {Component[]} blocks - a collection of story blocks (sjrk.storyTelling.block)
+     * @param {Object} savedStoryData - story model data loaded from autosave
+     * @param {Component} story - an instance of `sjrk.storyTelling.base.page.storyEdit`
+     * @param {Component} blockManager - the storyEditor's blockManager component
+     * @param {String} sourceName - the name of the Infusion change source for this update
+     */
+    sjrk.storyTelling.base.page.storyEdit.loadStoryContent = function (savedStoryData, story, blockManager, sourceName) {
+        story.applier.change("", savedStoryData, null, sourceName);
+
+        // build the blockUIs from the story content array
+        blockManager.createBlocksFromData(savedStoryData.content);
+    };
```

Comment:
It seems this whole function can be removed. Instead you could make the related invoker declaratively trigger the changeApplier. You could then have a modelListener, which could only listen to a specific source if need be, to call `blockManager.createBlocksFromData`.

Reply:
I agree with this change and partially implemented it, but I've run into trouble setting up the model listener.

I have tried:
1. setting up the model listener on the `storyEdit` grade (top-level, where the original `loadStoryContent` invoker is)
2. moving the model listener to the `storyEditor` component (mid-level, it owns the `blockManager` and `story` components)
3. moving the model listener to the `storyEditor`'s story component (the model being listened to)
4. moving the model listener to the `blockManager` (whose invoker is called by the listener) and setting its key (i.e. path) to `"{story}.model"`
5. moving the model listener to the `blockManager` and setting its key to `"{storyEditor}.story.model"`
6. moving the model listener to the `blockManager` and setting the `segs` and `context` keys to `[]` and `"story"`, respectively

Solutions 1, 2 and 3 don't seem to be working due to timing issues. The `blockManager` component has a `createOnEvent` defined for it, and so when the `story` is being constructed, `blockManager` may not exist yet. That said, it should exist when `storyEdit` is being constructed, so this is surprising. Here is the error message in one of those cases, showing the definition of the model listener:
```
"Assertion failure - check console for more details: Error in listener record - could not resolve reference
{
    "func": "{storyEditor}.blockManager.createBlocksFromData",
    "args": ["content"],
    "includeSource": ["storyAutoload"],
    "namespace": "createEditorBlocksFromStoryData"
}
to a listener or firer. Did you miss out "events." when referring to an event firer?"
```

Solutions 3, 4 and 5 don't work either, though no errors are thrown and it fails silently.

Reply:
I'm having a bit of trouble understanding who should be responsible for which parts. It looks like this is a method on `sjrk.storyTelling.base.page.storyEdit` which modifies the model of the `storyEditor` subcomponent's `story` subcomponent's model. It then makes a call to the `storyEditor`'s `blockManager`'s `createBlocksFromData` method. I'm not entirely clear on the distinction between `storyEdit` and `storyEditor`. I'm guessing the former is the page, and the latter is just the portion where stories are written. Is that correct?

Reply:
I think I have this working with the attached changes. However, I'm still not sure of the factoring. 
[SJRK-289.patch.zip](https://github.com/fluid-project/sjrk-story-telling/files/5517903/SJRK-289.patch.zip)

Reply:
That's correct: `storyEdit` is the page which includes the `storyEditor`, `storyViewer` and a couple other `ui` grade subcomponents

`storyEditor` is the `ui` component which has the story (which in turn has the data), a templateManager which renders everything, the blockManager (which manages the blockUi components which each in turn have a template manager & block)

This patch worked, thanks! I've opted to use the previous event source value (`storyAutoLoad`) which was already present, rather than the new one in your path (`loadStoryContent`, named after the invoker that initiates the change). The reason for this is that the former is the source for the changes triggered in `sjrk.storyTelling.base.page.storyEdit.createNewStoryOnServer`, and we don't need/want to call `createBlocksFromData` at this point.